So far I have a clean/new `scitran` instance running on my machine with `./live.sh run`. I also have the reaper from this PR running with:

```
./meeg_file_reaper.py -s 1 -u http://localhost:9001/scitran ~/Desktop/test/
```

And the output is:

```
2015-08-12 11:12:24      reaper.meeg:INFO Monitoring acquisition directory /home/larsoner/Desktop/test/
2015-08-12 11:12:24           reaper:INFO initializing MEEGFileReaper...
2015-08-12 11:12:24           reaper:INFO loaded       1 items from persistence file, 1 not reaped
2015-08-12 11:12:24           reaper:INFO monitoring   /home/larsoner/Desktop/test/eric_non_space/maddox_ross/20150810 [2015-08-10 15:09:25, 1 files, 38M]
2015-08-12 11:12:24           reaper:INFO monitoring   1 items, 1 not reaped
2015-08-12 11:12:25           reaper:INFO reap queue   item 1 of 1
2015-08-12 11:12:25      reaper.meeg:INFO reaping      /home/larsoner/Desktop/test/eric_non_space/maddox_ross/20150810 [2015-08-10 15:09:25, 1 files, 38M]
2015-08-12 11:12:25      reaper.meeg:INFO reaped       /home/larsoner/Desktop/test/eric_non_space/maddox_ross/20150810 (1 images) in 0.3s
2015-08-12 11:12:25      reaper.meeg:INFO compressing  /home/larsoner/Desktop/test/eric_non_space/maddox_ross/20150810
2015-08-12 11:12:29           reaper:INFO hashing      reap.tgz
2015-08-12 11:12:29           reaper:INFO uploading    reap.tgz [27M] to http://localhost:9001/scitran
2015-08-12 11:12:29           reaper:ERRO error        reap.tgz: ('Connection aborted.', error(32, 'Broken pipe'))
```

I suspect I have a badly formatted URI, but what should it be? I tried `https` but I got an un-caught error:

```
...
2015-08-12 11:26:55           reaper:INFO uploading    reap.tgz [27M] to https://localhost:9001/scitran
Traceback (most recent call last):
  File "./meeg_file_reaper.py", line 124, in <module>
    main()
  File "./meeg_file_reaper.py", line 120, in main
    reaper.main(MEEGFileReaper, positional_args, optional_args)
  File "/home/larsoner/python/scitran/code/reaper/scitran/reaper/reaper.py", line 324, in main
    reaper.run()
  File "/home/larsoner/python/scitran/code/reaper/scitran/reaper/reaper.py", line 190, in run
    if self.upload(tempdir):
  File "/home/larsoner/python/scitran/code/reaper/scitran/reaper/reaper.py", line 252, in upload
    success = self.http_upload(filename, filepath, digest, uri)
  File "/home/larsoner/python/scitran/code/reaper/scitran/reaper/reaper.py", line 275, in http_upload
    r = requests.put(uri, data=fd, headers=headers)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/api.py", line 110, in put
    return request('put', url, data=data, **kwargs)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/api.py", line 49, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/sessions.py", line 461, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 544, in urlopen
    body=body, headers=headers)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 344, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=conn.timeout)
  File "/home/larsoner/.local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 314, in _raise_timeout
    if 'timed out' in str(err) or 'did not complete (read)' in str(err):  # Python 2.6
TypeError: __str__ returned non-string (type SysCallError)
```

Is this documented somewhere? I'm starting to add it to `README.md` in this PR.

Also, what is the suggested folder structure for development? Should this `reaper` repo be checked out into the folder from the `scitran` repo `scitran/code/`?

--- a/README.md
+++ b/README.md
@@ -11,9 +11,10 @@ pip install -r requirements.txt .
 ### Reaping
 
 ```
-reaper/pfile_reaper.py <path>
-reaper/dicom_reaper.py <host> <port> <return port> reaper <scanner AET>
-reaper/dicom_file_reaper.py <path>
+./reaper/pfile_reaper.py <path>
+./reaper/dicom_reaper.py <host> <port> <return port> reaper <scanner AET>
+./reaper/dicom_file_reaper.py <path>
+./reaper/meeg_file_reaper.py <path>
 
 ```
 
@@ -23,4 +24,11 @@ reaper/dicom_file_reaper.py <path>
 ```
 findscu --verbose -S -aet reaper -aec <scanner AET> -k QueryRetrieveLevel="STUDY" -k StudyDate="" <host> <port>
 findscu --verbose -S -aet reaper -aec <scanner AET> -k QueryRetrieveLevel="SERIES" -k StudyDate="" <host> <port>
+```
+
+
+### Development
+
+```
+./reaper/meeg_file_reaper.py -s 1 -u localhost <path>
 ```

--- a/README.md
+++ b/README.md
@@ -30,5 +30,5 @@ findscu --verbose -S -aet reaper -aec <scanner AET> -k QueryRetrieveLevel="SERIE
 ### Development
 
 ```
-./reaper/meeg_file_reaper.py -s 1 -u localhost <path>
+./reaper/meeg_file_reaper.py -s 1 -u https://localhost:8443/api/reaper?secret=change-me <path>
 ```

--- a/scitran/reaper/meeg_file_reaper.py
+++ b/scitran/reaper/meeg_file_reaper.py
@@ -1,0 +1,124 @@
+#!/usr/bin/env python
+#
+# @author:  Eric Larson
+
+"""
+M/EEG file reaper
+"""
+
+import logging
+log = logging.getLogger('reaper.meeg')
+log.setLevel(logging.INFO)
+
+import os
+import sys
+import shutil
+import datetime
+import warnings
+
+import reaper
+
+
+class MEEGFileReaper(reaper.Reaper):
+
+    def __init__(self, options):
+        self.path = options.get('path')
+        if not os.path.exists(self.path):
+            os.mkdir(self.path)
+            log.info(self.path + ' dost not exist ==> creating...')
+        elif not os.path.isdir(self.path):
+            log.error('path argument must be a directory')
+            sys.exit(1)
+        else:
+            log.info('Monitoring acquisition directory %s' % self.path)
+        self.destructive = options.get('destructive')
+        super(MEEGFileReaper, self).__init__(
+            self.path.strip('/').replace('/', '_'), options)
+
+    def state_str(self, _id, state):
+        return '%s [%s, %d files, %s]' % (
+            _id,
+            state['mod_time'].strftime(reaper.DATE_FORMAT),
+            state['file_cnt'],
+            reaper.hrsize(state['size']),
+        )
+
+    def instrument_query(self):
+        i_state = {}
+        for dirpath, dirnames, filenames in os.walk(self.path):
+            if os.path.basename(dirpath).startswith('.'):
+                continue  # ignore dotdirectories
+            if os.path.samefile(dirpath, self.path):
+                continue  # ignore files at top-level of self.path
+            if len(dirnames) > 0 or len(filenames) == 0:
+                continue
+            try:
+                mtime = os.path.getmtime(dirpath)
+                state = {
+                    'mod_time': datetime.datetime.utcfromtimestamp(mtime),
+                    'file_cnt': len(filenames),
+                    'size': sum([os.path.getsize(os.path.join(dirpath, fn))
+                                 for fn in filenames]),
+                }
+            except:
+                pass
+            else:
+                i_state[dirpath] = reaper.ReaperItem(state, path=dirpath)
+        return i_state
+
+    def reap(self, _id, item, tempdir):
+        reap_start = datetime.datetime.utcnow()
+        log.info('reaping      %s' % self.state_str(_id, item['state']))
+        reap_cnt = 0
+        reap_path = os.path.join(tempdir, 'reap')
+        os.mkdir(reap_path)
+        for fn in os.listdir(item['path']):
+            fp = os.path.join(item['path'], fn)
+            try:
+                # catch warnings here because some recordings have a
+                # shielding mode that will throw a warning on read
+                import mne
+                with warnings.catch_warnings(record=True):
+                    mne.io.read_raw_fif(fp, allow_maxshield=True)
+            except Exception:
+                pass
+            else:
+                reap_cnt += 1
+                shutil.copyfile(fp, reap_path + '/' + fn)
+        t = (datetime.datetime.utcnow() - reap_start).total_seconds()
+        log.info('reaped       %s (%d images) in %.1fs' % (_id, reap_cnt, t))
+
+        # project is the third-level dir name (project/subject/data/x_raw.fif)
+        project = \
+            os.path.split(os.path.split(os.path.split(item['path'])[0])[0])[1]
+
+        metadata = {
+            'filetype': 'FIF',
+            'timezone': self.timezone,
+            'header': {
+                'project': project,
+            },
+        }
+
+        log.info('compressing  %s' % _id)
+        reaper.create_archive(reap_path+'.tgz', reap_path,
+                              os.path.basename(reap_path), metadata,
+                              compresslevel=6)
+        shutil.rmtree(reap_path)
+        return True
+
+    def destroy(self, item):
+        shutil.rmtree(item['path'])
+
+
+def main():
+    positional_args = [
+        (('path',), dict(help='path to MEEG files')),
+    ]
+    optional_args = [
+    ]
+    reaper.main(MEEGFileReaper, positional_args, optional_args)
+
+
+if __name__ == '__main__':
+    main()

--- a/scitran/reaper/meeg_file_reaper.py
+++ b/scitran/reaper/meeg_file_reaper.py
@@ -101,7 +101,7 @@ class MEEGFileReaper(reaper.Reaper):
         }
 
         log.info('compressing  %s' % _id)
-        reaper.create_archive(reap_path+'.tgz', reap_path,
+        reaper.create_archive(reap_path+'.zip', reap_path,
                               os.path.basename(reap_path), metadata,
                               compresslevel=6)
         shutil.rmtree(reap_path)

--- a/scitran/reaper/reaper.py
+++ b/scitran/reaper/reaper.py
@@ -80,6 +80,10 @@ class Reaper(object):
         self.id_ = id_
         self.persitence_file = options.get('persistence_file')
         self.upload_uris = options.get('upload') or []
+        for uri in self.upload_uris:
+            if not self.get_upload_function(uri):
+                raise RuntimeError('Bad URI "%s", reaper will not function'
+                                   % uri)
         self.peripheral_data = dict(options.get('peripheral') or [])
         self.sleeptime = options.get('sleeptime') or SLEEPTIME
         self.graceperiod = datetime.timedelta(seconds=(options.get('graceperiod') or GRACEPERIOD))
@@ -238,21 +242,26 @@ class Reaper(object):
             for uri in self.upload_uris:
                 log.info('uploading    %s [%s] to %s' % (filename, hrsize(os.path.getsize(filepath)), uri))
                 start = datetime.datetime.utcnow()
-                if uri.startswith('http://') or uri.startswith('https://'):
-                    success = self.http_upload(filename, filepath, digest, uri)
-                elif uri.startswith('s3://'):
-                    success = self.s3_upload(filename, filepath, digest, uri)
-                elif uri.startswith('file://'):
-                    success = self.file_copy(filename, filepath, digest, uri)
-                else:
-                    log.error('unknown URI schem: %s' % uri)
-                    return False
+                success = self.get_upload_function(uri)(
+                    filename, filepath, digest, uri)
                 upload_duration = (datetime.datetime.utcnow() - start).total_seconds()
                 if success:
                     log.info('uploaded     %s [%s/s]' % (filename, hrsize(os.path.getsize(filepath)/upload_duration)))
                 else:
                     return False
         return True
+
+    def get_upload_function(self, uri):
+        """Helper to get an appropriate upload function based on protocol"""
+        if uri.startswith('http://') or uri.startswith('https://'):
+            return self.http_upload
+        elif uri.startswith('s3://'):
+            return self.s3_upload
+        elif uri.startswith('file://'):
+            return self.file_copy
+        else:
+            log.error('unknown URI schem: %s' % uri)
+            return False
 
     def http_upload(self, filename, filepath, digest, uri):
         headers = {

--- a/scitran/reaper/reaper.py
+++ b/scitran/reaper/reaper.py
@@ -245,10 +245,9 @@ class Reaper(object):
                 success = self.get_upload_function(uri)(
                     filename, filepath, digest, uri)
                 upload_duration = (datetime.datetime.utcnow() - start).total_seconds()
-                if success:
-                    log.info('uploaded     %s [%s/s]' % (filename, hrsize(os.path.getsize(filepath)/upload_duration)))
-                else:
+                if not success:
                     return False
+                log.info('uploaded     %s [%s/s]' % (filename, hrsize(os.path.getsize(filepath)/upload_duration)))
         return True
 
     def get_upload_function(self, uri):
@@ -261,7 +260,7 @@ class Reaper(object):
             return self.file_copy
         else:
             log.error('unknown URI schem: %s' % uri)
-            return False
+            return lambda w, x, y, z: False
 
     def http_upload(self, filename, filepath, digest, uri):
         headers = {

